File: frontend/src/lib/queries/trade-flow/decision-logs.ts
import { pool } from '@/lib/db';

interface DecisionLogExportRow {
  created_at: string;
  event_ts: string;
  event_type: string;
  decision_id: string | null;
  sl_event_id: string | null;
  fill_event_id: string | null;
  root_order_id: string | null;
  order_id: string | null;
  market_slug: string | null;
  asset: string | null;
  workflow: string | null;
  outcome: string | null;
  payload_json: unknown;
}

function csvField(value: string | number | null): string {
  if (value == null) return '';
  const text = String(value);
  if (!/[",\r\n]/.test(text)) return text;
  return `"${text.replaceAll('"', '""')}"`;
}

export async function buildDecisionLogsRawCsv(params: {
  userId: number;
  from?: string | null;
  to?: string | null;
}): Promise<string> {
  const queryParams: unknown[] = [params.userId];
  const clauses = ['o.user_id = $1'];

  if (params.from) {
    queryParams.push(params.from);
    clauses.push(`l.event_ts >= $${queryParams.length}::timestamptz`);
  }
  if (params.to) {
    queryParams.push(params.to);
    clauses.push(`l.event_ts <= $${queryParams.length}::timestamptz`);
  }

  const res = await pool.query<DecisionLogExportRow>(
    `SELECT
       l.created_at::text,
       l.event_ts::text,
       l.event_type,
       l.decision_id,
       l.sl_event_id,
       l.fill_event_id,
       l.root_order_id,
       l.order_id,
       l.market_slug,
       l.asset,
       l.workflow,
       l.outcome,
       l.payload AS payload_json
     FROM bot_decision_logs l
     JOIN trade_builder_orders o ON o.id::text = COALESCE(l.root_order_id, l.order_id)
     WHERE ${clauses.join(' AND ')}
     ORDER BY l.event_ts ASC, l.created_at ASC, l.id ASC`,
    queryParams
  );

  const headers = [
    'created_at',
    'event_ts',
    'event_type',
    'decision_id',
    'sl_event_id',
    'fill_event_id',
    'root_order_id',
    'order_id',
    'market_slug',
    'asset',
    'workflow',
    'outcome',
    'payload_json',
  ];
  const lines = [headers.map(csvField).join(',')];
  for (const row of res.rows) {
    lines.push(
      [
        row.created_at,
        row.event_ts,
        row.event_type,
        row.decision_id,
        row.sl_event_id,
        row.fill_event_id,
        row.root_order_id,
        row.order_id,
        row.market_slug,
        row.asset,
        row.workflow,
        row.outcome,
        JSON.stringify(row.payload_json ?? {}),
      ]
        .map((value) => csvField(value == null ? null : String(value)))
        .join(',')
    );
  }
  return `${lines.join('\n')}\n`;
}

File: crates/bot-infra/src/db/trade_builder/decision_logs.rs
use super::super::*;

impl PostgresRepository {
    pub async fn append_bot_decision_log(&self, input: &BotDecisionLogInput) -> Result<bool> {
        let result = sqlx::query(
            "INSERT INTO bot_decision_logs \
             (event_id, idempotency_key, schema_version, event_type, event_ts, decision_id, \
              sl_event_id, fill_event_id, market_slug, root_order_id, order_id, exchange_order_id, \
              parent_order_id, child_order_id, source_trade_id, flow_run_id, flow_definition_id, \
              pair_session_id, asset, workflow, outcome, outcome_token_id, opposite_token_id, payload) \
             VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, \
                     $17, $18, $19, $20, $21, $22, $23, $24) \
             ON CONFLICT DO NOTHING",
        )
        .bind(input.event_id)
        .bind(&input.idempotency_key)
        .bind(input.schema_version)
        .bind(&input.event_type)
        .bind(input.event_ts)
        .bind(&input.decision_id)
        .bind(&input.sl_event_id)
        .bind(&input.fill_event_id)
        .bind(&input.market_slug)
        .bind(&input.root_order_id)
        .bind(&input.order_id)
        .bind(&input.exchange_order_id)
        .bind(&input.parent_order_id)
        .bind(&input.child_order_id)
        .bind(&input.source_trade_id)
        .bind(&input.flow_run_id)
        .bind(&input.flow_definition_id)
        .bind(&input.pair_session_id)
        .bind(&input.asset)
        .bind(&input.workflow)
        .bind(&input.outcome)
        .bind(&input.outcome_token_id)
        .bind(&input.opposite_token_id)
        .bind(&input.payload)
        .execute(self.pool())
        .await?;
        Ok(result.rows_affected() > 0)
    }

    pub async fn list_bot_decision_logs_for_roots(
        &self,
        root_order_ids: &[String],
    ) -> Result<Vec<BotDecisionLogRecord>> {
        if root_order_ids.is_empty() {
            return Ok(Vec::new());
        }

        let rows = sqlx::query(
            "SELECT event_id, idempotency_key, schema_version, event_type, event_ts, created_at, \
                    decision_id, sl_event_id, fill_event_id, market_slug, root_order_id, order_id, \
                    exchange_order_id, parent_order_id, child_order_id, source_trade_id, flow_run_id, \
                    flow_definition_id, pair_session_id, asset, workflow, outcome, outcome_token_id, \
                    opposite_token_id, payload \
             FROM bot_decision_logs \
             WHERE root_order_id = ANY($1::text[]) \
             ORDER BY root_order_id ASC, event_ts ASC, created_at ASC, id ASC",
        )
        .bind(root_order_ids)
        .fetch_all(self.pool())
        .await?;

        Ok(rows.into_iter().map(map_bot_decision_log_row).collect())
    }
}

fn map_bot_decision_log_row(row: sqlx::postgres::PgRow) -> BotDecisionLogRecord {
    BotDecisionLogRecord {
        event_id: row.get("event_id"),
        idempotency_key: row.get("idempotency_key"),
        schema_version: row.get("schema_version"),
        event_type: row.get("event_type"),
        event_ts: row.get("event_ts"),
        created_at: row.get("created_at"),
        decision_id: row.get("decision_id"),
        sl_event_id: row.get("sl_event_id"),
        fill_event_id: row.get("fill_event_id"),
        market_slug: row.get("market_slug"),
        root_order_id: row.get("root_order_id"),
        order_id: row.get("order_id"),
        exchange_order_id: row.get("exchange_order_id"),
        parent_order_id: row.get("parent_order_id"),
        child_order_id: row.get("child_order_id"),
        source_trade_id: row.get("source_trade_id"),
        flow_run_id: row.get("flow_run_id"),
        flow_definition_id: row.get("flow_definition_id"),
        pair_session_id: row.get("pair_session_id"),
        asset: row.get("asset"),
        workflow: row.get("workflow"),
        outcome: row.get("outcome"),
        outcome_token_id: row.get("outcome_token_id"),
        opposite_token_id: row.get("opposite_token_id"),
        payload: row.get("payload"),
    }
}
